File: disk_agent/snapshot.py
from __future__ import annotations

import json
import os
import shutil
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Sequence, Tuple

from .models import DirectoryUsage, FilesystemUsage, PodmanUsage, Snapshot

SNAPSHOT_DIR = Path.home() / ".disk-agent" / "snapshots"
COMMAND_TIMEOUT_SECONDS = 120
TOP_DIRECTORY_DEPTH = 3


def _display_path(path: Path) -> str:
    home = Path.home()
    try:
        relative = path.resolve().relative_to(home.resolve())
    except (OSError, ValueError):
        return str(path)
    return "~" if str(relative) == "." else f"~/{relative}"


def _run(command: Sequence[str]) -> Tuple[str, str, int]:
    try:
        result = subprocess.run(
            command,
            text=True,
            capture_output=True,
            timeout=COMMAND_TIMEOUT_SECONDS,
            check=False,
            env={**os.environ, "LC_ALL": "C"},
        )
        return result.stdout, result.stderr, result.returncode
    except subprocess.TimeoutExpired as exc:
        return exc.stdout or "", f"command timed out after {COMMAND_TIMEOUT_SECONDS}s", 124
    except OSError as exc:
        return "", str(exc), 127


def _collect_filesystem() -> FilesystemUsage:
    stdout, stderr, code = _run(["df", "-B1", "--output=source,size,used,avail,pcent,target", "/"])
    lines = [line.split() for line in stdout.splitlines()[1:] if line.strip()]
    if code != 0 or not lines or len(lines[-1]) < 6:
        raise RuntimeError(f"df failed: {stderr.strip() or 'invalid output'}")
    filesystem, total, used, available, percent = lines[-1][:5]
    mountpoint = " ".join(lines[-1][5:])
    return FilesystemUsage(
        filesystem=filesystem,
        mountpoint=mountpoint,
        total_bytes=int(total),
        used_bytes=int(used),
        available_bytes=int(available),
        used_percent=int(percent.rstrip("%")),
    )


def _collect_du(path: Path, max_depth: int) -> Tuple[List[DirectoryUsage], List[str]]:
    if not path.exists():
        return [], [f"path not found: {_display_path(path)}"]
    stdout, stderr, code = _run(
        ["du", "-x", "-B1", f"--max-depth={max_depth}", str(path)]
    )
    values: List[DirectoryUsage] = []
    for line in stdout.splitlines():
        size, separator, raw_path = line.partition("\t")
        if not separator:
            continue
        try:
            values.append(DirectoryUsage(_display_path(Path(raw_path)), int(size)))
        except ValueError:
            continue
    warnings = []
    if code != 0 and stderr.strip():
        # du commonly returns 1 for unreadable paths. Keep useful partial data.
        warnings.append(f"du {_display_path(path)}: permission or read errors ignored")
    return values, warnings


def _parse_size(value: str) -> int:
    value = value.strip()
    if not value or value == "0B":
        return 0
    units = {"B": 1, "kB": 1000, "KB": 1000, "MB": 1000**2, "GB": 1000**3, "TB": 1000**4}
    for suffix in sorted(units, key=len, reverse=True):
        if value.endswith(suffix):
            try:
                return int(float(value[: -len(suffix)]) * units[suffix])
            except ValueError:
                return 0
    return 0


def _collect_podman() -> PodmanUsage:
    if shutil.which("podman") is None:
        return PodmanUsage(error="podman is not installed")
    stdout, stderr, code = _run(["podman", "system", "df", "--format", "{{json .}}"])
    if code != 0:
        return PodmanUsage(error=stderr.strip() or "podman system df failed")
    totals = {"Images": 0, "Containers": 0, "Local Volumes": 0}
    try:
        for line in stdout.splitlines():
            row = json.loads(line)
            kind = row.get("Type")
            if kind in totals:
                totals[kind] = _parse_size(str(row.get("Size", "0B")))
    except (json.JSONDecodeError, TypeError):
        return PodmanUsage(error="could not parse podman system df output")
    return PodmanUsage(
        available=True,
        images_bytes=totals["Images"],
        containers_bytes=totals["Containers"],
        volumes_bytes=totals["Local Volumes"],
    )


def collect_snapshot(now: datetime | None = None) -> Snapshot:
    now = now or datetime.now(timezone.utc).astimezone()
    home = Path.home()
    home_usage, warnings = _collect_du(home, 2)
    local_usage, local_warnings = _collect_du(home / ".local" / "share", 2)
    copilot_usage, copilot_warnings = _collect_du(home / ".copilot", 2)
    top_usage, top_warnings = _collect_du(home, TOP_DIRECTORY_DEPTH)
    warnings.extend(local_warnings + copilot_warnings + top_warnings)

    # Dedupe paths and retain the most precise observed value.
    by_path = {}
    for usage in top_usage + local_usage + copilot_usage:
        by_path[usage.path] = usage
    largest = sorted(by_path.values(), key=lambda item: item.bytes, reverse=True)
    largest = [item for item in largest if item.path != "~"][:100]

    return Snapshot(
        timestamp=now.isoformat(timespec="seconds"),
        filesystem=_collect_filesystem(),
        home_usage=home_usage,
        local_share_usage=local_usage,
        copilot_usage=copilot_usage,
        podman=_collect_podman(),
        largest_directories=largest,
        warnings=warnings,
    )


def save_snapshot(snapshot: Snapshot, directory: Path = SNAPSHOT_DIR) -> Path:
    directory.mkdir(parents=True, exist_ok=True)
    day = snapshot.timestamp[:10]
    destination = directory / f"{day}.json"
    temporary = destination.with_suffix(".json.tmp")
    with temporary.open("w", encoding="utf-8") as handle:
        json.dump(snapshot.to_dict(), handle, indent=2, sort_keys=True)
        handle.write("\n")
    temporary.replace(destination)
    return destination


def snapshot_command() -> str:
    snapshot = collect_snapshot()
    path = save_snapshot(snapshot)
    message = f"Snapshot stored: {path}"
    if snapshot.warnings:
        message += f"\nCompleted with {len(snapshot.warnings)} ignored warning(s)."
    return message

File: src/snapshot.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{anyhow, Result};
use chrono::{Local, SecondsFormat};

use crate::filesystem::{collect_du, collect_filesystem};
use crate::json::write_snapshot_pretty;
use crate::models::{DirectoryUsage, Snapshot};
use crate::paths;
use crate::podman::collect_podman;

pub const TOP_DIRECTORY_DEPTH: u8 = 3;

pub fn save_snapshot(snapshot: &Snapshot, directory: &Path) -> Result<PathBuf> {
    fs::create_dir_all(directory)?;
    let day = snapshot
        .timestamp
        .get(..10)
        .ok_or_else(|| anyhow!("snapshot timestamp is too short"))?;
    let destination = directory.join(format!("{day}.json"));
    let temporary = destination.with_extension("json.tmp");
    write_snapshot_pretty(&temporary, snapshot)?;
    fs::rename(&temporary, &destination)?;
    Ok(destination)
}

pub fn snapshot_command(verbose: bool) -> Result<String> {
    let snapshot = collect_snapshot()?;
    let path = save_snapshot(&snapshot, &paths::snapshot_dir()?)?;
    Ok(format_snapshot_message(&snapshot, &path, verbose))
}

fn format_snapshot_message(snapshot: &Snapshot, path: &Path, verbose: bool) -> String {
    let mut message = format!("Snapshot stored: {}", path.display());
    let warnings = if verbose {
        snapshot.warnings.iter().collect::<Vec<_>>()
    } else {
        snapshot
            .warnings
            .iter()
            .filter(|warning| !is_expected_warning(warning))
            .collect::<Vec<_>>()
    };
    if !warnings.is_empty() {
        message.push_str(&format!(
            "\nCompleted with {} ignored warning(s).",
            warnings.len()
        ));
        if verbose {
            message.push_str("\n\nIgnored warnings:");
            for warning in warnings {
                message.push_str(&format!("\n- {warning}"));
            }
        } else {
            message.push_str(
                "\nHint: re-run with `disk-agent snapshot --verbose` to view the ignored warnings.",
            );
        }
    }
    message
}

fn is_expected_warning(warning: &str) -> bool {
    warning.starts_with("du ") && warning.ends_with(": permission or read errors ignored")
}

pub fn collect_snapshot() -> Result<Snapshot> {
    let home = paths::home_dir()?;
    let (home_usage, mut warnings) = collect_du(&home, 2)?;
    let (local_share_usage, local_warnings) = collect_du(&home.join(".local").join("share"), 2)?;
    let (copilot_usage, copilot_warnings) = collect_du(&home.join(".copilot"), 2)?;
    let (top_usage, top_warnings) = collect_du(&home, TOP_DIRECTORY_DEPTH)?;
    warnings.extend(local_warnings);
    warnings.extend(copilot_warnings);
    warnings.extend(top_warnings);

    let largest_directories = largest_directories(&top_usage, &local_share_usage, &copilot_usage);

    Ok(Snapshot {
        timestamp: Local::now().to_rfc3339_opts(SecondsFormat::Secs, false),
        filesystem: collect_filesystem()?,
        home_usage,
        local_share_usage,
        copilot_usage,
        podman: collect_podman()?,
        largest_directories,
        warnings,
        schema_version: 1,
    })
}

fn largest_directories(
    top_usage: &[DirectoryUsage],
    local_share_usage: &[DirectoryUsage],
    copilot_usage: &[DirectoryUsage],
) -> Vec<DirectoryUsage> {
    let mut by_path = std::collections::HashMap::new();
    for usage in top_usage
        .iter()
        .chain(local_share_usage.iter())
        .chain(copilot_usage.iter())
    {
        by_path.insert(usage.path.clone(), usage.clone());
    }
    let mut largest = by_path.into_values().collect::<Vec<_>>();
    largest.sort_by(|left, right| right.bytes.cmp(&left.bytes));
    largest
        .into_iter()
        .filter(|usage| usage.path != "~")
        .take(100)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::format_snapshot_message;
    use crate::models::{FilesystemUsage, Snapshot};

    fn snapshot_with_warnings(warnings: Vec<&str>) -> Snapshot {
        Snapshot {
            timestamp: "2026-07-11T10:50:00+00:00".to_string(),
            filesystem: FilesystemUsage {
                filesystem: "/dev/vda".to_string(),
                mountpoint: "/".to_string(),
                total_bytes: 1000,
                used_bytes: 600,
                available_bytes: 400,
                used_percent: 60,
            },
            home_usage: Vec::new(),
            local_share_usage: Vec::new(),
            copilot_usage: Vec::new(),
            podman: Default::default(),
            largest_directories: Vec::new(),
            warnings: warnings.into_iter().map(str::to_string).collect(),
            schema_version: 1,
        }
    }

    #[test]
    fn default_snapshot_output_omits_warning_report_when_there_are_no_warnings() {
        let snapshot = snapshot_with_warnings(Vec::new());
        let output = format_snapshot_message(&snapshot, "/tmp/2026-07-11.json".as_ref(), false);

        assert_eq!(output, "Snapshot stored: /tmp/2026-07-11.json");
    }

    #[test]
    fn default_snapshot_output_omits_expected_warnings() {
        let snapshot = snapshot_with_warnings(vec![
            "du ~: permission or read errors ignored",
            "du ~/.local/share: permission or read errors ignored",
        ]);
        let output = format_snapshot_message(&snapshot, "/tmp/2026-07-11.json".as_ref(), false);

        assert_eq!(output, "Snapshot stored: /tmp/2026-07-11.json");
    }

    #[test]
    fn default_snapshot_output_reports_unexpected_warnings() {
        let snapshot = snapshot_with_warnings(vec![
            "du ~: permission or read errors ignored",
            "Path disappeared during scan: ~/gone",
        ]);
        let output = format_snapshot_message(&snapshot, "/tmp/2026-07-11.json".as_ref(), false);

        assert_eq!(
            output,
            "Snapshot stored: /tmp/2026-07-11.json\nCompleted with 1 ignored warning(s).\nHint: re-run with `disk-agent snapshot --verbose` to view the ignored warnings."
        );
    }

    #[test]
    fn verbose_snapshot_output_lists_ignored_warnings() {
        let snapshot = snapshot_with_warnings(vec![
            "du ~: permission or read errors ignored",
            "Path disappeared during scan: ~/gone",
        ]);
        let output = format_snapshot_message(&snapshot, "/tmp/2026-07-11.json".as_ref(), true);

        assert!(output.contains("Completed with 2 ignored warning(s)."));
        assert!(output.contains("Ignored warnings:\n- du ~: permission or read errors ignored"));
        assert!(output.contains("\n- Path disappeared during scan: ~/gone"));
    }
}
